Unify flow matching and score-based models

StarostinV opened the pull request:
## What does this PR do?

This unifies score-based models with flow matching under a common API. Furthermore, based on the following paper [https://arxiv.org/abs/2410.02217](https://arxiv.org/abs/2410.02217), one can use the vector field learnt with flow matching to calculate the time-dependent score and marginal distributions, as well as drift and diffusion functions for SDE-based sampling. This effectively enables the use of all score-based methods (SDE-based sampling, iid, gradient evaluation, map with score-based gradient, guidance, etc) with flow matching.

API of `FMPE` and `NPSE` didn't change, but they are now wrappers around the same class, `VectorFieldInference`, and only differ in the following:
- default `sample_with` value in `build_posterior` method (`sde` for NPSE and `ode` for FMPE),
- estimator builder (`FlowMatchingEstimator` and `ScoreBasedEstimator` that are subclasses of `ConditionalVectorFieldEstimator`),
- loss function and concrete SDE & ODE functions provided by the estimator.

Additionally, the ode solver is isolated into a separate API, which enables swapping ode backends. So far, only `zuko` backend is implemented.

## Does this close any issues?
#1440 and #1462

## Anything else we should know?

The following features have been implemented:

- [x] Add `ode_solvers` to be able to swap backends
- [x] Extend API in the  `ConditionalVectorFieldEstimator` to enable SDE 
- [x] Make `FlowMatchingEstimator` a subclass of  `ConditionalVectorFieldEstimator`
- [x] Implement score-based methods in `FlowMatchingEstimator`
- [x] Implement unified `VectorFieldPotential`, `VectorFieldPosterior`, `VectorFieldInference` classes
- [x] Update `FMPE` and `NPSE` classes
- [x] Update `score_fn_iid` to support new API
- [x] Remove tests that rely on old API and `DirectPosterior` for flow matching
- [x] Better documentation
- [x] Add tests for new functionality
- [x] Add benchmark tests
- [x] Remove the previous implementation of score-based methods

## Related issues

- This PR does not change the NN builders and does not unify NN architectures, which is another issue #1442 

## ✅ Checklist

Put an x in the boxes that apply. If you're unsure about any of them, no worries — just ask!

- [x] I have read and followed the [contribution guidelines](https://github.com/sbi-dev/sbi/blob/main/CONTRIBUTING.md)
- [x] I have added helpful comments to my code where needed
- [x] I have added tests for new functionality
- [ ] (If applicable) I have reported how long new tests run and marked them with pytest.mark.slow

**For reviewers:**

- [ ] I have reviewed every file
- [ ] All comments have been addressed

manuelgloeckler commented on sbi/samplers/ode_solvers/base.py:
```diff
@@ -0,0 +1,64 @@
+from zuko.lazy import LazyDistribution
+
+NeuralODEFuncType = Callable[[Tensor, Tensor, Tensor], Tensor]
+
```
This currently is not really used. You could add this is a `Protocol` i.e. see [here](https://github.com/StarostinV/sbi/blob/2a6f3016924843dca195460f56080b67c0090116/sbi/inference/potentials/base_potential.py#L85).

manuelgloeckler commented on sbi/samplers/ode_solvers/zuko_ode.py:
```diff
@@ -0,0 +1,85 @@
+
+from sbi.samplers.ode_solvers.base import NeuralODE, NeuralODEFuncType
+
+
```
Add a small docstringa

manuelgloeckler replied:
And description of arguments

manuelgloeckler commented on tests/embedding_net_test.py:
```diff
@@ -25,7 +25,7 @@
 
 @pytest.mark.mcmc
-@pytest.mark.parametrize("method", ["NPE", "NLE", "NRE", "FMPE"])
+@pytest.mark.parametrize("method", ["NPE", "NLE", "NRE"])
```
This might also be another issue as it wasnt there before. But FMPE and NPSE should be tested to at least "work" with embedding networks.

StarostinV replied:
I think this can be addressed by #1501 since it relies on network builders.

StarostinV replied:
I wrote down a list of TODOs in that PR, and it is also referenced in the comments to the code.

manuelgloeckler commented on tests/linearGaussian_fmpe_test.py:
```diff
@@ -15,60 +15,45 @@
+@pytest.mark.parametrize("num_dim", (2, 1))
+def test_c2st_fmpe_on_linearGaussian(num_dim: int):
     """Test whether fmpe infers well a simple example with available ground truth."""
 
```
Why is it necessary to remove uniform?

StarostinV replied:
The util get_normalization_uniform_prior expects DirectPosterior and calls log_prob with norm_posterior=True, which is not supported by VectorFieldPosterior (and its predecessor ScoreBasedPosterior).

manuelgloeckler reviewed (commented):
Great effort. Looks great. There is still some work on passing all slow tests. But @janfb this will be quite a big change so you can already have a look.

StarostinV commented:
I am currently making some fixes to add support for iid methods by flow matching. That will basically unify the tests for scores and flow matching. After that we can run all slow tests.

janfb commented on sbi/inference/posteriors/vector_field_posterior.py:
```diff
@@ -65,25 +68,28 @@ def __init__(
         )
         # Set the potential function type.
-        self.potential_fn: PosteriorScoreBasedPotential = potential_fn
+        self.potential_fn: PosteriorVectorFieldBasedPotential = potential_fn
```
the naming might now be a bit confusing here. before, we called this `PosteriorScoreBasedPotential` because the potential is based on the "posterior score". 
I think now we could just call it `VectorFieldBasedPotential`?

StarostinV replied:
Absolutely, I will rename it.

janfb commented on sbi/inference/posteriors/vector_field_posterior.py:
```diff
@@ -144,31 +150,27 @@ def sample(
-                f"`sample_with` is no longer supported. You have to rerun "
-                f"`.build_posterior(sample_with={sample_with}).`"
-            )
+        if sample_with is None:
```
agreed! 

but removing the deprecating error makes sense because we will eventually remove `build_posterior` and allow again interacting with `posterior.sample` directly. 

So, just add a ValueError like Manuel suggested.

janfb commented on sbi/inference/posteriors/vector_field_posterior.py:
```diff
@@ -277,10 +282,9 @@ def sample_via_ode(
 
-        flow = self.potential_fn.get_continuous_normalizing_flow(
-            condition=self.potential_fn.x_o
+        samples = self.potential_fn.neural_ode(self.potential_fn.x_o).sample(
+            torch.Size((num_samples,))
```
👍

janfb commented on sbi/inference/potentials/score_fn_iid.py:
```diff
@@ -76,12 +78,19 @@ def __init__(
 
-        self.score_estimator = score_estimator.to(device).eval()
+        if not vector_field_estimator.SCORE_DEFINED:
+            raise ValueError("Score is not defined for this vector field estimator.")
+        if not vector_field_estimator.MARGINALS_DEFINED:
+            raise ValueError(
+                "Marginals are not defined for this vector field estimator."
```
good to have this. can you please add a bit more detail in the error message, e.g., what could be the reason? 
I guess one reason could be that the flow matching based training does not support it, e.g, when using optimal transport flows?

janfb commented on sbi/inference/potentials/score_fn_iid.py:
```diff
@@ -109,7 +118,7 @@ def __call__(
 class FNPEScoreFunction(IIDScoreFunction):
```
I am a bit concerned about this name because it is so close to `FMPE`. can we think of a different name here? `FactorizedNPE`?

janfb reviewed (commented):
This is amazing. I am posting this 25% review, will continue later on the train.

StarostinV commented:
@janfb @manuelgloeckler I have some good news: the iid methods work great with flow matching! However, so far the metrics in (slow iid) tests are not so great, and that is because of the neural network architecture. I trained a model independently using a very simple architecture and c2st is much better for iid sampling with flow matching. That is another PR though.

StarostinV commented:
Btw I have addressed most of your comments in the recent commits (I haven't implemented Protocol for NeuralODEFunc but maybe I will look into this later).

janfb commented on sbi/samplers/ode_solvers/base.py:
```diff
@@ -0,0 +1,100 @@
+                    - input (Tensor): The input state tensor :math:`\theta_t`
+                    - condition (Tensor): The conditioning tensor :math:`x_o`
+                    - time (Tensor): The time parameter tensor :math:`t`
+            net: The neural network that is used by the function :math:`f`.
+                This is never called explicitly by the NeuralODE class,
+                but is used to track the parameters of the neural network.
```
👍 please add this docstring to the `build_neural_ode` fun as well.

StarostinV replied:
Done.

janfb commented on tests/density_estimator_test.py:
```diff
@@ -253,13 +253,12 @@ def test_correctness_of_density_estimator_log_prob(
     "input_event_shape", ((1,), pytest.param((2,), marks=pytest.mark.slow))
 )
-@pytest.mark.parametrize("condition_event_shape", ((1,), (2,)))
+@pytest.mark.parametrize("condition_event_shape", ((1,), (7,)))
```
I recall that I changed this to `(2,)` to speed up the tests. is there a reason to use `(7,)`?

janfb commented on tests/linearGaussian_vector_field_test.py:
```diff
@@ -0,0 +1,585 @@
+    )
+
+
+# TODO: This should be unified with NPSE when the network builders are unified.
```
please add as TODO in the #1501 PR.

janfb commented on tests/linearGaussian_vector_field_test.py:
```diff
@@ -0,0 +1,585 @@
+    )
+
+
+# TODO: Currently, c2st is too high for FMPE (e.g., > 3 number of observations),
+# so some tests are skipped so far. It is unclear at the moment whether this is due to
+# some issues with the implementation, the nn architecture, or it is simply the case
+# with flow matching (compositional flow matching hasn't been tested yet elsewhere).
```
can you please still add them explicitly as test cases here, add a `skip`, and create an issue, thanks! 🙏

janfb commented on tests/linearGaussian_vector_field_test.py:
```diff
@@ -0,0 +1,585 @@
+# ------------------------------------------------------------------------------
+
+
+# TODO: Need to add NPSE when the network builders are unified, but anyway
+# this will only work after implementing additional methods for vector fields,
+# so it is skipped for now.
```
please add a TODO in that PR, or an issue.

janfb commented on sbi/samplers/ode_solvers/__init__.py:
```diff
@@ -0,0 +1,51 @@
+]
+
+
+def build_neural_ode(
+    f: NeuralODEFuncType,
+    net: nn.Module,
+    mean_base: Tensor,
+    std_base: Tensor,
+    backend: str = "zuko",
+    t_min: float = 0.0,
+    t_max: float = 1.0,
+    **kwargs,
+) -> NeuralODE:
```
we usually do not have functions in `__init__.py`. move this to `net_builders.py`?

StarostinV replied:
Moved to ode_builder.py

janfb commented on sbi/samplers/ode_solvers/__init__.py:
```diff
@@ -0,0 +1,51 @@
+]
+
+
+def build_neural_ode(
+    f: NeuralODEFuncType,
+    net: nn.Module,
+    mean_base: Tensor,
+    std_base: Tensor,
+    backend: str = "zuko",
+    t_min: float = 0.0,
+    t_max: float = 1.0,
+    **kwargs,
+) -> NeuralODE:
```
we usually do not have functions in `__init__.py`. move this to `net_builders.py`?

janfb commented on sbi/neural_nets/estimators/score_estimator.py:
```diff
@@ -338,6 +365,23 @@ def _set_weight_fn(self, weight_fn: Union[str, Callable]):
+        score = self.forward(input=input, condition=condition, time=t)
+        f = self.drift_fn(input, t)
+        g = self.diffusion_fn(input, t)
+        v = f - 0.5 * g**2 * score
```
where does this equation come from? can you please add a short comment or pointer to a paper / eqn?

StarostinV replied:
I added a reference into the docstring

janfb commented on sbi/neural_nets/estimators/flowmatching_estimator.py:
```diff
@@ -20,147 +16,337 @@ class VectorFieldNet(nn.Module, ABC):
-        zscore_transform_input: Optional[Transform] = None,
-        num_freqs: int = 3,
+        embedding_net: Optional[nn.Module] = None,
+        num_freqs: int = 3,  # This is ignored!
         noise_scale: float = 1e-3,
+        zscore_transform_input=None,  # This is ignored!
```
I think it could be fine to remove these un-used args here. or were they user-facing? anyway, I think it's also fine to change the API here.

StarostinV replied:
This is already changed in #1501 (I couldn't do that myself without breaking the tests or editing net builders)

janfb commented on sbi/neural_nets/estimators/flowmatching_estimator.py:
```diff
@@ -20,147 +16,337 @@ class VectorFieldNet(nn.Module, ABC):
+    def loss(
+        self, input: Tensor, condition: Tensor, times: Optional[Tensor] = None, **kwargs
+    ) -> Tensor:
+        r"""Return the loss for training the density estimator. More precisely,
```
please add a new line after the first sentence.

StarostinV replied:
fixed.

janfb commented on sbi/neural_nets/estimators/flowmatching_estimator.py:
```diff
@@ -20,147 +16,337 @@ class VectorFieldNet(nn.Module, ABC):
+        notation is changed to match the standard SBI notation: :math:`\theta_0 = x_0`
+        and :math:`\theta_1 = x_1`).
+
+        The loss is computed as the mean squared error between the vector field:
+
+        .. math::
+            L(\theta_0, \theta_1, t, x_o) = \| v(\theta_t, t; x_o) -
+            (\theta_1 - \theta_0) \|^2
+
+            where
+
+            .. math::
+                \theta_1 \sim p_{base}
+                \theta_t = t \cdot \theta_1 + (1 - t) \cdot \theta_0
+
+            Additionally, the small noise :math:`\sigma_{min}` is added to the
+            vector field as per [2]_ to address numerical issues at small times.
```
is this new? and related to the Singh et al. paper?

StarostinV replied:
I've updated the docstring for the loss, but the loss itself stays the same and comes from the Lipman paper [2] (and Liu [1]). The Singh paper only shows how to use this standard rectified flow to sample via SDE.

janfb commented on sbi/neural_nets/estimators/flowmatching_estimator.py:
```diff
@@ -20,147 +16,337 @@ class VectorFieldNet(nn.Module, ABC):
-            NormalizingFlow: flow with log_prob and sample methods via probability ODEs.
+            Score function of the vector field estimator.
         """
+        # TODO: derive taking into account the noise sigma_min added to the vector field
```
how relevant is this? worth another issue?

StarostinV replied:
Actually, it would be great to fix it. I can open another issue if I don't manage to derive this myself over the weekend and fix here - it should add a small change to the score and the marginals, which currently assume sigma_min=0. It might accumulate over the integration steps.

StarostinV replied:
It is fixed now - the analytical score and marginals incorporate sigma_min.

janfb commented on sbi/neural_nets/estimators/flowmatching_estimator.py:
```diff
@@ -20,147 +16,337 @@ class VectorFieldNet(nn.Module, ABC):
         """
+        # TODO: derive taking into account the noise sigma_min added to the vector field
+        v = self(input, condition, t)
+        score = (-(1 - t) * v - input) / torch.maximum(t, torch.tensor(1e-6))
```
do we need to take into account the `device` when creating the `Tensor` here?

StarostinV replied:
Yes!

StarostinV replied:
In general, should I add some gpu tests?

manuelgloeckler replied:
Yes, that would be great. FMPE and NPSE should be added [here](https://github.com/sbi-dev/sbi/blob/d9ac6336bc0f6c9ed27bee6adf114b0febae208f/tests/inference_on_device_test.py#L98).

But this actually wasn't tested before well. So we could also move it to another issue/PR.

StarostinV replied:
I've created an issue for that #1530

janfb commented on sbi/neural_nets/estimators/flowmatching_estimator.py:
```diff
@@ -20,147 +16,337 @@ class VectorFieldNet(nn.Module, ABC):
+            Drift function at a given time.
+        """
+        # TODO: derive taking into account the noise sigma_min added to the vector field
+        return -input / torch.maximum(1 - times, torch.tensor(1e-6))
```
maybe add device handling.

janfb reviewed (commented):
Wow, a lot of work went into this, well done @StarostinV ! 👏 

I added tons of comments, naive questions and links to other PRs, sorry about that. but we need to make sure this integrates well with the other changes. 

There are a lot of TODOs in line-comments in the code. Ideally, we resolve all of them berfore merging this, e.g., by creating issues or adding them as explicit tasks in the "sister-PR" #1501.

StarostinV commented on sbi/inference/posteriors/vector_field_posterior.py:
```diff
@@ -229,9 +236,10 @@ def _sample_via_diffusion(
 
+        # TODO: the time schedule should be provided by the vector field estimator.
         if ts is None:
-            t_max = self.score_estimator.t_max
-            t_min = self.score_estimator.t_min
+            t_max = self.vector_field_estimator.t_max
+            t_min = self.vector_field_estimator.t_min
```
The issue with time schedules for SDE samplers #1437 is unrelated to the changes made in this PR, but I wanted to highlight this place when somebody starts fixing it. There is already a corresponding PR #1481, but I haven't cheched yet how it is fixed there. 

The problem is that some SDE samplers can benefit from nonlinear time discretization (e.g., VE). Currently, the time steps are generated via linspace in the posterior, and in my view that responsibility should be passed to the Diffuser to handle times internally based on some methods (e.g., time_schedule) provided by the VectorFieldEstimator.

StarostinV commented on sbi/inference/potentials/vector_field_potential.py:
```diff
@@ -51,32 +56,49 @@ def score_estimator_based_potential(
 
         Args:
-            score_estimator: The neural network modelling the score.
+            vector_field_estimator: The neural network modelling the score.
```
Thanks, spot on! I actually see that some other docstrings in the vector field classes should be also fixed - I will go through them. I also noticed that some docstrings in the package use latex for math notation and others use the standard :math: . Is latex supported and what is the preferred option? So far I've been using :math:

StarostinV replied:
I've updated the docstrings here in some other places now.

StarostinV commented on sbi/neural_nets/estimators/flowmatching_estimator.py:
```diff
@@ -20,147 +16,337 @@ class VectorFieldNet(nn.Module, ABC):
-            transform=transform,
-            base=DiagNormal(self.zeros, self.ones).expand(condition.shape[:-1]),
-        )
+        The drift function is calculated based on [3]_ (see Equation 7):
+
+        .. math::
+            f(t) = - \theta_t / (1 - t)
+
+        The drift function :math:`f(t)` and diffusion function :math:`\g(t)`
+        enable SDE sampling:
+
+        .. math::
+            d\theta_t = [f(t) - g(t)^2 \nabla_{\theta_t} \log p(\theta_t | x_o)]dt
+            + \g(t) dW_t
+
```
Yes, this comes from the Singh et al. paper along with the score and marginals for flow matching. In general, they show there are infinitely many SDEs one could construct, but so far I have implemented the most basic one.

StarostinV commented on sbi/inference/potentials/vector_field_potential.py:
```diff
@@ -101,46 +121,56 @@ def set_x(
         self,
         theta: Tensor,
-        track_gradients: bool = True,
+        track_gradients: bool = False,
```
I had a discussion with @manuelgloeckler and we agreed that it should be set to False by default for vector fields. In short, it obviously gives a significant overhead and is only required in very special usecases. So it makes sense to explicitly provide this argument in these rare cases.

StarostinV commented on sbi/inference/potentials/vector_field_potential.py:
```diff
@@ -51,32 +56,49 @@ def score_estimator_based_potential(
         self.iid_method = iid_method
         self.iid_params = iid_params
+
+        neural_ode_kwargs = neural_ode_kwargs or {}
+        self.neural_ode = build_neural_ode(
+            self.vector_field_estimator.ode_fn,
+            self.vector_field_estimator.net,
+            self.vector_field_estimator.mean_base,
+            self.vector_field_estimator.std_base,
+            backend=neural_ode_backend,
+            t_min=self.vector_field_estimator.t_min,
+            t_max=self.vector_field_estimator.t_max,
+            **neural_ode_kwargs,
+        )
```
Yes, this creates the newly introduced NeuralODE that is used for log_prob calculation and sampling via ODE. Previously this was done in less flexible way in the method rebuild_flow, and it wasn't possible to add other ODE backends.

StarostinV commented on sbi/inference/trainers/fmpe/fmpe.py:
```diff
@@ -53,286 +41,57 @@ def __init__(
+
+        For FMPE, the posterior distribution that is returned here implements
+        the following functionality over the raw neural density estimator:
+            - correct the calculation of the log probability such that
+              it compensates for the leakage.
```
VectorFieldPotential.log_prob sets log prob of samples outside the prior support to -inf, but doesn't change log prob of samples within the support. I will write it down more clearly.

StarostinV replied:
Fixed.

StarostinV commented on sbi/neural_nets/estimators/vector_field_estimator.py:
```diff
@@ -0,0 +1,225 @@
+        Raises:
+            NotImplementedError: This method should be implemented by sub-classes.
+        """
+        raise NotImplementedError("This method should be implemented by sub-classes.")
+
```
The reason why all these methods are not abstract is that we want to support vector field estimators that don't have analytical score, marginals, etc. One can define such estimator, set e.g. SCORE_DEFINED = False, and leave score method unimplemented without getting error by abc. However, forward and ode_fn should be implemented for every vector field estimator - forward returns the learned field and the ode_fn is defined for all the score-based methods and flow matching.

StarostinV commented on sbi/inference/trainers/npse/vector_field_inference.py:
```diff
@@ -0,0 +1,552 @@
+from sbi.utils.torchutils import assert_all_finite
+
+
+class VectorFieldInference(NeuralInference):
```
It is resolved now.

StarostinV commented on sbi/inference/potentials/vector_field_potential.py:
```diff
@@ -51,32 +56,49 @@ def score_estimator_based_potential(
+        vector_field_estimator: ConditionalVectorFieldEstimator,
         prior: Optional[Distribution],
         x_o: Optional[Tensor] = None,
         iid_method: str = "auto_gauss",
```
Nice idea! I added Literal for iid_method and sde_type.

StarostinV commented on sbi/inference/potentials/vector_field_potential.py:
```diff
@@ -51,32 +56,49 @@ def score_estimator_based_potential(
             iid_params: Parameters for the iid method, for arguments see
                 `IIDScoreFunction`.
             device: The device on which to evaluate the potential.
+            neural_ode_backend: The backend to use for the neural ODE.
```
Done.

StarostinV commented on sbi/inference/trainers/fmpe/fmpe.py:
```diff
@@ -53,286 +41,57 @@ def __init__(
+    ) -> VectorFieldPosterior:
+        r"""Build posterior from the flow matching estimator. Note that
+        this is the same as the NPSE posterior, but the sample_with method
+        is set to "ode" by default.
```
Fixed.

StarostinV commented on sbi/inference/trainers/npse/vector_field_inference.py:
```diff
@@ -0,0 +1,552 @@
+    def __init__(
+        self,
+        prior: Optional[Distribution] = None,
+        vector_field_estimator: Union[str, Callable] = "mlp",
```
I added a protocol for the vector field estimator builder. I keep the API in NPSE and FMPE that call this builder "score_estimator" and "density_estimator" correspondingly, but that seems misleading, since it is not the estimator but the builder. Should I open an issue for renaming this?

StarostinV commented on sbi/inference/trainers/npse/vector_field_inference.py:
```diff
@@ -0,0 +1,552 @@
+                string.
+            summary_writer: Tensorboard summary writer.
+            show_progress_bars: Whether to show progress bars during training.
+            kwargs: Additional keyword arguments.
```
Fixed.

StarostinV commented on sbi/inference/trainers/npse/vector_field_inference.py:
```diff
@@ -0,0 +1,552 @@
+            x: Simulation outputs.
+            proposal: The distribution that the parameters $\theta$ were sampled from.
+                Pass `None` if the parameters were sampled from the prior. If not
+                `None`, it will trigger a different loss-function.
```
Yes! Fixed.

StarostinV commented on sbi/inference/trainers/npse/vector_field_inference.py:
```diff
@@ -0,0 +1,552 @@
+        # Check for problematic z-scoring
+        warn_if_zscoring_changes_data(x)
+
+        npe_msg_on_invalid_x(num_nans, num_infs, exclude_invalid_x, "Single-round NPE")
```
Done.

StarostinV commented on sbi/inference/trainers/npse/vector_field_inference.py:
```diff
@@ -0,0 +1,552 @@
+        # Otherwise, infer it from the device of the net parameters.
+        else:
+            # TODO: Add protocol for checking if the score estimator has forward and
+            # loss methods with the correct signature.
```
@manuelgloeckler I am not sure why a protocol would be needed here (this comment comes from the NPSE implementation since I haven't changed the train implementation) - we specify that we expect ConditionalVectorFieldEstimator. I removed this comment and ignore type and pyright doesn't have any problems with it.

StarostinV commented on sbi/inference/trainers/npse/vector_field_inference.py:
```diff
@@ -0,0 +1,552 @@
+         is implemented, i.e., no proposal correction is applied for later rounds.
+
+        The loss is the negative log prob. Irrespective of the round or SNPE method
+        (A, B, or C), it can be weighted with a calibration kernel.
```
Fixed.

StarostinV commented on sbi/neural_nets/estimators/vector_field_estimator.py:
```diff
@@ -0,0 +1,225 @@
+from sbi.neural_nets.estimators.base import ConditionalEstimator
+
+
+class ConditionalVectorFieldEstimator(ConditionalEstimator):
```
I think I had some circular imports at the beginning but then it changed, so I moved it back to base.py.

StarostinV commented on sbi/neural_nets/estimators/vector_field_estimator.py:
```diff
@@ -0,0 +1,225 @@
+        Raises:
+            NotImplementedError: This method should be implemented by sub-classes.
+        """
+
```
Right, it is removed for abstract methods.

StarostinV commented on sbi/inference/potentials/vector_field_potential.py:
```diff
@@ -101,46 +121,56 @@ def set_x(
         super().set_x(x_o, x_is_iid)
         self.iid_method = iid_method
         self.iid_params = iid_params
         # NOTE: Once IID potential evaluation is supported. This needs to be adapted.
         # See #1450.
```
The corresponding PR #1508 is blocked by this PR and is supposed to incorporate it into the new vector field classes.

StarostinV commented on sbi/inference/trainers/npse/vector_field_inference.py:
```diff
@@ -0,0 +1,552 @@
+from sbi.utils.torchutils import assert_all_finite
+
+
+class VectorFieldInference(NeuralInference):
```
There is a docstring in the VectorFieldInference, do you mean class-level docstring? I think no inference class has that

janfb commented:
> @janfb @manuelgloeckler I have some good news: the iid methods work great with flow matching! However, so far the metrics in (slow iid) tests are not so great, and that is because of the neural network architecture. I trained a model independently using a very simple architecture and c2st is much better for iid sampling with flow matching. That is another PR though.

that's great! but what is the difference in the NN architecture between ours and your simple one?

janfb reviewed (changes requested):
Thanks again @StarostinV for another round of big effort! This looks almost done now 👍 

Added a couple of minor comments and questions.

StarostinV commented on sbi/inference/posteriors/vector_field_posterior.py:
```diff
@@ -204,6 +219,7 @@ def _sample_via_diffusion(
         show_progress_bars: bool = True,
     ) -> Tensor:
         r"""Return samples from posterior distribution $p(\theta|x)$.
+        Note that this method can be unsupported for some vector field estimators.
```
The only problem I have with such concrete comments is that I haven't worked with OT flow matching myself, so I don't know whether it is true that we can't derive an analytical score there.

janfb replied:
agreed, this is actually just a hypothesis that it might not work with OT. so just leave out the OT example but say that it just depends on whether the corresponding score is defined?

StarostinV replied:
Added

StarostinV commented on sbi/inference/potentials/vector_field_potential.py:
```diff
@@ -51,48 +57,70 @@ def score_estimator_based_potential(
+        the vector field estimator does not define the score (SCORE_DEFINED = False),
+        the gradient is not available and an error is raised.
+
+        Note that the potential function is not defined for the iid setting yet.
```
Yes, but it is blocked by this PR. I will add a comment into the code to refer to that PR.

StarostinV commented on tests/linearGaussian_vector_field_test.py:
```diff
@@ -0,0 +1,586 @@
+
+from .test_utils import check_c2st, get_dkl_gaussian_prior
+
+# ------------------------------------------------------------------------------
+# ------------------------------- FAST TESTS -----------------------------------
+# ------------------------------------------------------------------------------
```
Yes, this is rather "not marked as slow since it is important/agreed to test on every run"  than "fast tests" :). It is quite slow indeed. I preserved the logic of the tests, but in my view, it can be made lighter. For instance, a single sde type would be enough in the "fast" tests, mark other types as slow, and probably check some fast things for these sde types in the "fast" setting. I can open an issue for that (this PR is already quite large and blocks others).

janfb replied:
no need for an issue, we will do it when we merge #1461 or in the downstream refactoring.

StarostinV commented on sbi/neural_nets/estimators/base.py:
```diff
@@ -223,45 +223,224 @@ def sample_and_log_prob(
+
+    # -------------------------- SDE METHODS --------------------------
+
+    def score(self, input: Tensor, condition: Tensor, t: Tensor) -> Tensor:
+        r"""Time-dependent score function
+
+        .. math::
+            s(t, \theta_t; x_o) = \nabla_{\theta_t} \log p(\theta_t | x_o)
+
+        Args:
+            input: Input parameters :math:`\theta_t`.
+            condition: Conditioning variable :math:`x_o`.
+            t: Time.
+
+        Raises:
+            NotImplementedError: Score is not implemented for this estimator.
+        """
+        raise NotImplementedError("Score is not implemented for this estimator.")
```
Exactly, this is the idea.

StarostinV commented on sbi/inference/trainers/npse/vector_field_inference.py:
```diff
@@ -0,0 +1,600 @@
+from sbi.utils.torchutils import assert_all_finite
+
+
+class VectorFieldEstimatorBuilder(Protocol):
+    """Protocol for building a vector field estimator from data."""
+
+    def __call__(self, theta: Tensor, x: Tensor) -> ConditionalVectorFieldEstimator:
+        """Build a vector field estimator from theta and x, which mainly
+        inform the shape of the input and the condition to the neural network.
+        Generally, it can also be used to z-score the data, but not in the case
+        of vector field estimators.
+
+        Args:
+            theta: Parameter sets.
+            x: Simulation outputs.
+
+        Returns:
+            Vector field estimator.
+        """
+        ...
```
I replied there with a possible solution that doesn't take away the information about the concrete expected estimator.

StarostinV commented on sbi/inference/posteriors/vector_field_posterior.py:
```diff
@@ -27,35 +29,38 @@
 
     Log probabilities are obtained by calling the potential function, which in turn uses
-    zuko probabilistic ODEs to compute the log-probability.
+    ODE to compute the log-probability.
```
Fixed

StarostinV commented on sbi/neural_nets/estimators/score_estimator.py:
```diff
@@ -8,7 +8,9 @@
 
-from sbi.neural_nets.estimators.base import ConditionalVectorFieldEstimator
+from sbi.neural_nets.estimators.base import (
+    ConditionalVectorFieldEstimator,
```
Fixed here and in all the other files

StarostinV commented on tests/linearGaussian_vector_field_test.py:
```diff
@@ -0,0 +1,586 @@
+    elif vector_field_type == "fmpe":
+        inference = FMPE(prior=None)
+    else:
+        raise ValueError(f"Invalid vector field type: {vector_field_type}")
```
Done (and I ran the test)

StarostinV commented:
> > @janfb @manuelgloeckler I have some good news: the iid methods work great with flow matching! However, so far the metrics in (slow iid) tests are not so great, and that is because of the neural network architecture. I trained a model independently using a very simple architecture and c2st is much better for iid sampling with flow matching. That is another PR though.
> 
> that's great! but what is the difference in the NN architecture between ours and your simple one?

I haven't looked into the current architecture, but we made some quick tests during the hackathon and it was clear that the architecture for flow matching performs substantially worse than the simple one below. However, it could also be because of the number of parameters. It will become clear after unifying net builders for scores and flows. 

I just use some MLP with skip connections and time embeddings for tests, but I wouldn't advertise it since it could be improved in many ways :) Maybe I just put it here for the reference:

```python
class SimpleNet(VectorFieldNet):
    def __init__(
            self,
            in_dim: int = 2,
            condition_dim: int = 2,
            out_dim: int = 2, 
            hid_dim: int = 256, 
            time_emb_dim: int = 16,
            num_blocks: int = 3,
        ):
        super().__init__()
        self.time_embedding = SinusoidalTimeEmbedding(time_emb_dim)

        in_dims = [in_dim + time_emb_dim + condition_dim] + [hid_dim] * (num_blocks - 1)
        out_dims = [hid_dim] * (num_blocks - 1) + [out_dim]

        self.net = nn.Sequential(
            *[
                ResidualBlock(in_dim, out_dim, hid_dim=hid_dim) 
                for in_dim, out_dim in zip(in_dims, out_dims)
            ],
        )

    def forward(self, theta: torch.Tensor, x: torch.Tensor, t: torch.Tensor) -> torch.Tensor:
        x = torch.cat([theta, x, self.time_embedding(t)], dim=-1)
        x = self.net(x)
        return x


class ResidualBlock(nn.Module):
    def __init__(self, in_dim: int, out_dim: int, hid_dim: int = 128):
        super().__init__()
        self.net = nn.Sequential(
            nn.Linear(in_dim, hid_dim),
            nn.LeakyReLU(),
            nn.LayerNorm(hid_dim),
            nn.Linear(hid_dim, out_dim),
        )

        self.residual = nn.Linear(in_dim, out_dim)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        return self.net(x) + self.residual(x)
```

janfb reviewed (approved):
Heroic effort @StarostinV 🏅 🚀 

looks all good now, thanks for your patients with all my comments! 🙏 
looking forward to seeing this in action.
